kv/kvserver: TestReplicateQueueDecommissioningNonVoters failed

kv/kvserver.TestReplicateQueueDecommissioningNonVoters [failed](https://tanzanite.cluster.engflow.com/invocations/default/604a374f-0eb7-479b-8406-e604c5b70608?testReportRun=26&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [715628abd134abfd2c0d966f9b7220a6715cc299](https://github.com/cockroachdb/cockroach/commits/715628abd134abfd2c0d966f9b7220a6715cc299):


```
=== RUN   TestReplicateQueueDecommissioningNonVoters
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestReplicateQueueDecommissioningNonVoters3181076198
    test_log_scope.go:81: use -show-logs to present logs inline
    replicate_queue_test.go:831: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicateQueueDecommissioningNonVoters3181076198
--- FAIL: TestReplicateQueueDecommissioningNonVoters (45.89s)
=== RUN   TestReplicateQueueDecommissioningNonVoters/remove
    replicate_queue_test.go:823: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:823
        	Error:      	"1" is not greater than or equal to "2"
        	Test:       	TestReplicateQueueDecommissioningNonVoters/remove
        	Messages:   	expected replica removals to increase by at least 2
    --- FAIL: TestReplicateQueueDecommissioningNonVoters/remove (30.34s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=26</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueDecommissioningNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35905

kv/kvserver.TestReplicateQueueDecommissioningNonVoters [failed](https://tanzanite.cluster.engflow.com/invocations/default/7997c18f-6e7a-43c9-89c6-b37de47e69ca?testReportRun=12&testReportShard=17&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [41da9eccc216413c09cc66ad4c6bfa5f92c506e3](https://github.com/cockroachdb/cockroach/commits/41da9eccc216413c09cc66ad4c6bfa5f92c506e3):


```
=== RUN   TestReplicateQueueDecommissioningNonVoters
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestReplicateQueueDecommissioningNonVoters1807123972
    test_log_scope.go:81: use -show-logs to present logs inline
    replicate_queue_test.go:830: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicateQueueDecommissioningNonVoters1807123972
--- FAIL: TestReplicateQueueDecommissioningNonVoters (40.30s)
=== RUN   TestReplicateQueueDecommissioningNonVoters/remove
    replicate_queue_test.go:822: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:822
        	Error:      	"1" is not greater than or equal to "2"
        	Test:       	TestReplicateQueueDecommissioningNonVoters/remove
        	Messages:   	expected replica removals to increase by at least 2
    --- FAIL: TestReplicateQueueDecommissioningNonVoters/remove (25.48s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=17</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueDecommissioningNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -796,9 +796,31 @@ func TestReplicateQueueDecommissioningNonVoters(t *testing.T) {
 		require.NoError(t,
 			tc.Server(0).Decommission(ctx, livenesspb.MembershipStatus_DECOMMISSIONING, nonVoterNodeIDs))
 
+		testutils.SucceedsSoon(t, func() error {
+			// Ensure that the leaseholder store notices the decommissioning nodes
+			// before we re-enable the replicate queue. This is necessary because the
+			// replicate queue might otherwise race with the gossip update, removing
+			// the non-voters without noticing they are decommissioning, failing the
+			// RemoveDecommissioningNonVoterReplicaCount assertion below. See
+			// #115750.
+			repl, err := store.GetReplica(scratchRange.RangeID)
+			if err != nil {
+				return err
+			}
+			if decomRepls := store.GetStoreConfig().StorePool.DecommissioningReplicas(
+				repl.Desc().Replicas().Descriptors()); len(decomRepls) < 2 {
+				return errors.Errorf(
+					"expected 2 decommissioning replicas, found %d [%v]",
+					len(decomRepls), decomRepls)
+			}
+			return nil
+		})
+
 		// At this point, we know that we have an over-replicated range with
-		// non-voters on nodes that are marked as decommissioning. So turn the
-		// replicateQueue on and ensure that these redundant non-voters are removed.
+		// non-voters on nodes that are marked as decommissioning, and that the
+		// leaseholder store has received the gossip update which changes the
+		// non-voter node status to decommissioning. So turn the replicateQueue on
+		// and ensure that these redundant non-voters are removed.
 		tc.ToggleReplicateQueues(true)
 		require.Eventually(t, func() bool {
 			ok, err := checkReplicaCount(ctx, tc, &scratchRange, 1 /* voterCount */, 0 /* nonVoterCount */)

